Shared element transition for predictive back

maRci002 opened the pull request:
This PR adds a shared element transition for predictive back navigation, based on the specification found here: https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#shared-element-transition.

A new transition builder added: `PredictiveBackPageSharedElementTransitionsBuilder`, which constructs the shared element transition version.

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.light,
        pageTransitionsTheme: const PageTransitionsTheme(
          builders: {
            TargetPlatform.android: PredictiveBackPageSharedElementTransitionsBuilder(),
          },
        ),
      ),
      home: const FirstScreen(),
    );
  }
}

class FirstScreen extends StatelessWidget {
  const FirstScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('First Screen'),
      ),
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            Navigator.of(context).push(
              MaterialPageRoute(builder: (_) => const SecondScreen()),
            );
          },
          child: const Text('Go to Second Screen'),
        ),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  const SecondScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Theme(
      data: ThemeData(brightness: Brightness.dark),
      child: Scaffold(
        appBar: AppBar(
          title: const Text('Second Screen'),
        ),
        body: const Center(
          child: Text('Hello, Predictive back!'),
        ),
      ),
    );
  }
}
```

</details>

[predictive_back_shared_element.webm](https://github.com/user-attachments/assets/1194cf09-c64f-4b1f-aa3f-10feb0ef12bd)

Partial fix for #153577

maRci002 committed:
Add Shared element transition for predictive back

maRci002 commented:
cc @justinmc I have opened this PR as [we discussed](https://github.com/flutter/flutter/issues/153577#issuecomment-2332686003). I will add tests and documentation once the PR looks good.

maRci002 commented:
I think adding a new `PageTransitionBuilder` alongside `PredictiveBackPageTransitionsBuilder` would be better than using a builder parameter. For instance, the border radius could be disabled more easily by setting it to 0.

Piinks reviewed (commented):
Hi @maRci002 thanks for contributing! It looks like there are some failing checks, can you take a look?

justinmc commented on packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart:
```diff
@@ -37,6 +37,47 @@ class PredictiveBackPageTransitionsBuilder extends PageTransitionsBuilder {
+        // pop gesture. Otherwise, for things like button presses or other
+        // programmatic navigation, fall back to ZoomPageTransitionsBuilder.
+        if (route.popGestureInProgress) {
+          return _PredictiveBackPageFullScreenTransition(
```
I think PredictiveBackPageTransitionBuilder should probably do the shared element transition, since that seems to be the default on native Android? Then we would have a separate PredictiveBackFullScreenPageTransitionBuilder for that transition. Or we could even delete it... Do we know if the full screen transition exists on native Android? Like can I choose to use it when writing a native Android app? It's hard to tell since everything is in beta still.

justinmc reviewed (commented):
Thanks for jumping on this!  Sorry I missed the notification.

Two slight inconsistencies I notice when I try the native Settings app on the latest beta:

1. The previous route sits slightly offscreen to the left and also moves vertically when you drag vertically. I guess that's probably not easily doable. Possibly with @MitchellGoodwin's recent work in https://github.com/flutter/flutter/pull/150031.

2. After releasing the gesture, the outgoing route fades out on native, but on your branch I see it scale and then suddenly disappear.

Can you take a look and see if you can get closer to native on either of those 2 points? Otherwise this is a huge improvement and I'm on board with landing this approach.

https://github.com/user-attachments/assets/cdb7c19d-4bdd-41a8-a825-c8027fc1905a

Piinks commented:
@maRci002 do you plan to continue working on this PR?

maRci002 commented:
Yes, I plan to continue working on this PR. I will soon make one of my phones operational to see the native behavior accurately. I will place this PR in draft until then.

maRci002 commented:
On Android 14, the Settings app uses the `Full Screen Surfaces` animation, while on Android 15, the Settings app uses the `Shared Element` transition. More about these concepts: [GitHub Issue Comment](https://github.com/flutter/flutter/issues/153577#issuecomment-2311200825).

Here is the predictive back animation on Android 15 (_with both `Transition animation scale` and `Animator duration scale` set to 10x in the Developer options_):

https://github.com/user-attachments/assets/108709bc-ea8c-4b55-ac7a-40952a381c3d

> 2\. After releasing the gesture, the outgoing route fades out on native, but on your branch I see it scale and then suddenly disappear.

The only problem is that we are driving the route via [`_PredictiveBackGestureDetector`](https://github.com/flutter/flutter/blob/603104015dd692ea3403755b55d07813d5cf8965/packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart#L132). At this point, the page transition is nearly 100%, so I can't rely on the fade-out starting at 80%, because the user might change their mind and cancel the predictive back gesture. This means the `_PredictiveBackGestureDetector` should provide a phase to let the builder know whether it's in the commit phase. I think the actual transition should handle its own animation when the commit starts.

justinmc commented:
@maRci002 Can you re-upload that video? It doesn't ever start playing for me.

maRci002 committed:
Update Shared Element transition

maRci002 commented:
> @maRci002 Can you re-upload that video? It doesn't ever start playing for me.

https://github.com/user-attachments/assets/b3118ec1-7690-4630-a190-678f8523a063

Here is default back button behaviour:

https://github.com/user-attachments/assets/b9083ec6-00ea-42a2-90aa-59a7c47c754b

> 1. The previous route sits slightly offscreen to the left and also moves vertically when you drag vertically. I guess that's probably not easily doable. Possibly with @MitchellGoodwin's recent work in [Allow mixing route transitions in one app. #150031](https://github.com/flutter/flutter/pull/150031).

When `route.popGestureInProgress == false`, we fall back to `ZoomPageTransitionsBuilder`, which might be replaced by #142352. This means the previous route will animate offscreen by default. For reference, see the attached video about Android 15 back button behavior.
However, I can use #150031; the previous route should also be scaled/transitioned, which means I need actual predictive back events. The problem is that predictive back events from `WidgetsBinding._backGestureObserver` are sent only to the [active observer](https://github.com/flutter/flutter/blob/509cb43e9d98c608d5e7bf83f32816a578b24490/packages/flutter/lib/src/widgets/binding.dart#L871).

Here is a pseudo code about prev route problem:

```dart
class PredictiveBackPageSharedElementTransitionsBuilder extends PageTransitionsBuilder {
  const PredictiveBackPageSharedElementTransitionsBuilder();

  @override
  Widget delegatedTransition(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, bool allowSnapshotting, Widget? child) {
    // Problems:
    // We don't know Route (topmost)
    // We cannot listen Predictive back events since WidgetsBinding send only to the active listener which is the _PredictiveBackGestureDetector inside buildTransitions method
  }

  @override
  Widget buildTransitions<T>() {
    return _PredictiveBackGestureDetector<T>(
      route: route,
      builder: () {
        if (route.popGestureInProgress) {
          return _PredictiveBackPageSharedElementTransition();
        }

        return const ZoomPageTransitionsBuilder().buildTransitions();
      },
    );
  }
}
```

maRci002 commented:
> 2\. After releasing the gesture, the outgoing route fades out on native, but on your branch I see it scale and then suddenly disappear.

I made some changes and set [`timeDilation`](https://api.flutter.dev/flutter/scheduler/timeDilation.html) to 2.0, but unfortunately, my emulator is a bit slow.


https://github.com/user-attachments/assets/51cd0f70-0f32-4074-b954-3915d250a388

justinmc reviewed (commented):
I'm able to see your videos on my Mac but not on my Linux machine. Strange but I guess that is doable :shrug: .

Thank you for all of your assessment. I'm on board with everything you said. So in the latest Android the default page transition is the Shared Axis transition (https://github.com/flutter/flutter/issues/142352, which is being worked on), and when doing predictive back it uses the Shared Element transition (which this PR implements).

So we are on track to match the latest version of native Android. PredictiveBackPageTransitionBuilder will be updated to use your _PredictiveBackPageSharedElementTransition and fall back to the upcoming Shared Axis transition.

### Problems for the incoming route when swiping back
I think we could merge this PR without support for animating the incoming route. Your latest video you posted looks quite good without it. But let's talk through those problems.

#### We don't know the topmost route
Is that necessary? See for example @MitchellGoodwin's new PR that implements modal bottom sheet using delegatedTransition. Could you do something similar or am I misunderstanding?

https://github.com/flutter/flutter/pull/157568/files#diff-46d8c09c856d9697ebf30eb9e88b0eab83a0e3ca439711968544927bc9ec4daaR107

#### We can't listen to predictive back events

 > The problem is that predictive back events from WidgetsBinding._backGestureObserver are sent only to the [active observer](https://github.com/flutter/flutter/blob/509cb43e9d98c608d5e7bf83f32816a578b24490/packages/flutter/lib/src/widgets/binding.dart#L871).

This pattern of notifying the first observer keeps coming back to haunt me. I should have tried harder to move away from it for predictive back.

What if we change it to notify all observers? I think that's the way that this pattern was meant to be used. However, we still need a return value, so it would have to be something like this:

```dart
  WidgetsBindingObserver? _backGestureObserver;

  Future<bool> _handleStartBackGesture(Map<String?, Object?> arguments) {
    _backGestureObserver = null;
    final PredictiveBackEvent backEvent = PredictiveBackEvent.fromMap(arguments);
    bool handled = false;
    for (final WidgetsBindingObserver observer in List<WidgetsBindingObserver>.of(_observers)) {
      if (observer.handleStartBackGesture(backEvent)) {
        _backGestureObserver = observer;
        handled = true;
      }
    }
    return Future<bool>.value(handled);
  }
```

maRci002 committed:
Use delegatedTransition

justinmc commented:
@maRci002 How did it go when using delegatedTransition? I tried the PR locally and it didn't seem to move the incoming route at all.

maRci002 commented:
Apologies for not addressing this pull request recently; I've been quite swamped. I do plan to continue working on it, especially now that the [M3 page transition](https://github.com/flutter/flutter/pull/158881) has been implemented. However, if time is of the essence, please feel free to take over and continue this PR.

justinmc commented:
@maRci002 Thanks for following up. I'm winding down for the holidays here, so I will check on this PR again after the new year.

Piinks commented:
Update: Caught up with Justin last week regarding this change, he plans to pick this up to finish.

TheLastFlame commented:
I've been working on implementing this here: https://github.com/TheLastFlame/flutter/tree/shared-elements-predictive-back

It allows you to work with Hero, dialogs, and bottomsheets. However, there are still a few unresolved issues in the page transition implementation. 

I'd be happy to join the discussion.

TheLastFlame commented:
I was starting to make a demo for demoing and testing at the time:

https://github.com/user-attachments/assets/78f7a2f9-a603-4b80-997b-65704866b1be

justinmc commented:
@TheLastFlame Nice work! Is this how dialogs etc. are supposed to behave with predictive back according to the Material spec?

I had an old PR for OpenContainer animations kind of like Hero: https://github.com/flutter/packages/pull/6321

justinmc commented:
I'm assigning myself as I plan to finish this PR. @maRci002 I plan to push some updates while keeping your original commits.

justinmc committed:
Formatter

justinmc committed:
WIP exploring how to use delegatedTransition in the predictive back transitions

TheLastFlame commented:
@justinmc 

The spec doesn't focus specifically on dialogs and other modal elements, but that's how it works in Android 15/16. 

https://github.com/user-attachments/assets/7afeaa09-c76c-420c-9fef-3003d51eb1ea

I made a generic builder that makes it easy to add custom predicative animations to anything. I assume it should go into the final pr with this feature, as the ability to flexibly customize the predictive animation for any custom elements is really very cool.

I think we should implement it in dialog, bottomsheet and drawers builders by default

TheLastFlame commented:
In addition, one of the main problems with distributing predictive animation on Android is that by default it only works with standard page transitions. And most apps use their own custom animations.

So I added a parentTransitionsBuilder parameter that allows you to make any custom transition predicative. The only thing is, I haven't been able to make it work together with the CupertinoPageTransitionBuilder back gesture yet.

TheLastFlame commented:
I remembered what else is missing to match native Android behavior. 
The screen rounding should match the actual rounding of the device.
Currently flutter doesn't have an api to get this information.

We can of course leave a constant rounding, which in general will look more or less fine on most devices, but on some devices there will be a really strong difference with native.

maRci002 commented:
> I'm assigning myself as I plan to finish this PR. @maRci002 I plan to push some updates while keeping your original commits.

Thanks @justinmc for taking this on! I'll keep an eye on the PR but unfortunately can't dedicate more time to it. [`FadeForwardsPageTransitionsBuilder`](https://github.com/flutter/flutter/pull/158881) and [`Mixing Route Transitions`](https://github.com/flutter/flutter/pull/150031) are since released and they could be really helpful here.

justinmc commented:
@TheLastFlame Thanks for the info! I've created an issue about dialogs: https://github.com/flutter/flutter/issues/166608. Does it also affect bottom sheets and drawers as well on native?

I agree the screen rounding will have to be a constant middle ground for now, unfortunately.

justinmc committed:
Analyzer fixes

justinmc committed:
Cleanup, example working with delegatedTransition

justinmc committed:
Merge branch 'master' into predictive_back-shared_element_transition

justinmc committed:
Merge branch 'master' into predictive_back-shared_element_transition

justinmc committed:
Some adjustments to the shared element transition

As eyeballed by me on my phone.

justinmc committed:
Clean up the durations

justinmc committed:
Reorder methods and opacity to animation

justinmc committed:
The animation duration doesnt seem to depend on gesture progression to me.

justinmc committed:
Convert scale to an animation

justinmc committed:
WIP cleaning up animations and calculated values

justinmc committed:
Fix x commit animation

justinmc committed:
Offset animation instead of two double animations

justinmc committed:
Properly manage inheritedwidget dependency for xAnimation

justinmc committed:
_commitPositionAnimation also managers inheritedwidget properly

justinmc committed:
Constant cleanup and inheritedmodel

justinmc committed:
Finished cleaning up sharedelementtransition

justinmc committed:
Analyzer

justinmc committed:
Analyzer

justinmc committed:
Fullscreen animation primary cleanup

justinmc committed:
Analyzer

justinmc committed:
Clean up publicly exposed API

justinmc committed:
Docs analyzer

justinmc committed:
Docs polish

justinmc committed:
Improve constant names

justinmc committed:
Update widget name in tests

justinmc committed:
Clean up secondary animation:

justinmc committed:
Default transition on top

justinmc committed:
Border radius doesn't seem to animate

justinmc committed:
Listener was unused

justinmc committed:
Bring back the border radius animation

justinmc committed:
Fix leak

justinmc committed:
Clean up

justinmc committed:
Test both sharedelement and fullscreen transitions

justinmc committed:
WIP tracking down double back bug

justinmc committed:
WIP Possible solution to double back bug

justinmc committed:
Cleanup

justinmc committed:
Test that double back works.

justinmc commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -1858,16 +1858,6 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
     if (!animation!.isCompleted) {
       return false;
     }
-    // If we're being popped into, we also cannot be swiped until the pop above
-    // it completes. This translates to our secondary animation being
-    // dismissed.
-    if (!secondaryAnimation!.isDismissed) {
-      return false;
-    }
-    // If we're in a gesture already, we cannot start another.
-    if (popGestureInProgress) {
-      return false;
-    }
```
These seemed to be an incorrect assumption. In my testing I wanted to do two back gestures quickly back-to-back, and this code prevented that from working.

I added a test that does two backs quickly now.

TODO: See if any other tests are broken by this change..

justinmc commented:
@chunhtai Here's a quick summary of the changes in this PR to help with review.

Native Android has two predictive back gestures: [shared element](https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#shared-element-transition) and [fullscreen](https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#full-screen-surfaces). Previously full screen was the default (API 34), but now shared element is the default (API 35).

Before this PR, Flutter had only one predictive back page transition, PredictiveBackPageTransitionsBuilder, which implemented the full screen transition. After this PR, PredictiveBackPageTransitionsBuilder implements the shared element transition, and the full screen transition is in PredictiveBackSharedElementPageTransitionsBuilder.

This PR also makes improvements to the full screen transition's fidelity.

This PR does not implement delegatedTransition for these two transitions. That is tracked in https://github.com/flutter/flutter/issues/153577#issuecomment-2791066016.

chunhtai commented on packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart:
```diff
@@ -159,130 +265,363 @@ class _PredictiveBackGestureDetectorState extends State<_PredictiveBackGestureDe
+    _mergedAnimations = Listenable.merge(<Listenable>[widget.animation, _commitAnimation]);
+    _opacityAnimation = _opacityTween.animate(_commitAnimation);
+    _scaleAnimation = _scaleTween.animate(widget.animation);
+    _scaleCommitAnimation = Tween<double>(begin: _lastScale, end: 1.0).animate(_commitAnimation);
```
we only really need this when phase is commit right? otherwise, we would just throw it away later.

chunhtai commented on packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart:
```diff
@@ -159,130 +265,363 @@ class _PredictiveBackGestureDetectorState extends State<_PredictiveBackGestureDe
+  late final Listenable _mergedAnimations;
+  late final Animation<double> _opacityAnimation;
+  late final Animation<double> _scaleAnimation;
+  late Animation<double> _scaleCommitAnimation;
```
consider using a proxyAnimation to abstract away widget.animation and commit animation so that you don't have to recreate the animation every time phase changes

chunhtai reviewed (commented):
I hope we can abstract away different phase if possible. right now the code is kinda hard to read...

justinmc committed:
Simplify scaleCommitAnimation

justinmc committed:
scale abstracted to a single animation

justinmc committed:
updateAnimations method

justinmc committed:
WIP positionanimation, but the dragged y coordinate can't be an animation, so I'll need to change it.

maRci002 commented:
My concern is that `_PredictiveBackGestureDetectorState` is calling [`TransitionRoute.handleCommitBackGesture()`](https://github.com/flutter/flutter/blob/7a4b25601468943c8620984a63061286777b4050/packages/flutter/lib/src/widgets/routes.dart#L589), which is responsible for closing the current route:

```dart
@override
void handleCommitBackGesture() {
  phase = _PredictiveBackPhase.commit;

  widget.route.handleCommitBackGesture(); // here
  startBackEvent = currentBackEvent = null;
}
```

The problem is that [`TransitionRoute.handleCommitBackGesture()`](https://github.com/flutter/flutter/blob/7a4b25601468943c8620984a63061286777b4050/packages/flutter/lib/src/widgets/routes.dart#L589) contains hardcoded values. For instance:

```dart
final int droppedPageBackAnimationTime = ui.lerpDouble(0, 800, _controller!.value)!.floor();
```

It interpolates the duration up to 800ms based on the current progress (`_controller!.value`). This approach was acceptable for `FullPageTransition`, where the animation duration depended on how far the gesture progressed.

However, for shared element predictive back, this is no longer relevant. Regardless of whether the progress is at 60%, 80%, or 100%, the Android OS sends a `commitBackGesture` signal (_when user releases his/her finger_), and the resulting back animation should always last `_kCommitMilliseconds`.

That’s why `_PredictiveBackSharedElementPageTransitionState` has its own `_commitController`, which in my opinion should manage the logic for closing the route. This implies that `TransitionRoute` should not implement `PredictiveBackRoute`. Instead, it should simply expose its `controller`—similar to how `_CupertinoBackGestureController` does:

https://github.com/flutter/flutter/blob/7a4b25601468943c8620984a63061286777b4050/packages/flutter/lib/src/cupertino/route.dart#L199-L211

TheLastFlame commented:
I was a little disappointed in this pr after Justin decided to limit it to non delegatedTransition work.
As it is, I don't see much difference from the state it was in last year.
I'll wait until this pr is complete to try to do the delegatedTransition option on a new base.

TheLastFlame commented:
I was a little disappointed in this pr after Justin decided to limit it to non delegatedTransition work.
I'm not sure if the decision to abstract it isn't a mistake that will cause strong changes to need to be made afterward
I'll wait until this pr is complete to try to do the delegatedTransition option on a new base.

justinmc committed:
positionAnimation

justinmc commented:
@maRci002 Good call, we shouldn't have any of that visual logic in TransitionRoute. I will try to refactor it out.

@TheLastFlame My motivation is just to get this animation shipped quickly and in smaller pieces rather than one massive PR. It looks like the delegatedTransition work should be mostly separate from the work in this PR... I do want to avoid any breaking changes. Let me know if you see anything with the current PR that would need to be broken later. And I'm happy to help with code review if you want to open a PR for the delegatedTransition work afterwards!

justinmc committed:
WIP Clean up TransitionRoute

There should be no logic in there related to the predictive back
transition. I've removed it, but now I still need to clean up the
predictive back transition to make sure that it works correctly. Still
getting some jumpiness in the x position on commit. Border radius was
also messed up, but I fixed that.

justinmc committed:
Fixed x stutter.

justinmc committed:
Do border radius with _continuousAnimation

justinmc committed:
Port scale to _continuousAnimation, and fix _continuousAnimation post-commnit direction

justinmc committed:
Refactor away commitAnimation.

justinmc committed:
Backup is FadeForwards transition, not Zoom any more.

Also, speed up the commit part of the animation to match.

justinmc committed:
Now the back button transition is correct (fadeforwards)

justinmc committed:
Update existing curves to easeInOutCubicEmphasized.

This is what FadeForwards uses. There's still some discussion as to
whether or not that's the right curve, because FadeForwards still seems
too slow for me. It seems to animate very fast to very close to
completion, and then slow down. Maybe I'll change that curve in the
future.

justinmc committed:
Merge branch 'master' into predictive_back-shared_element_transition

Getting the PR that made the incoming route tappable during the back
transition.

justinmc committed:
Cleanup

justinmc committed:
Should stick with the old curve for Y.

justinmc committed:
_lastDrag Offset instead of 2 doubles

justinmc committed:
Cleanup.

The existing way is the best way to get the last used animation value.
At the time I need it, the animation has already been reset, so I do
have to save it somewhere.

And no radius is needed for the incoming route when pushed because the
corners are not exposed, so any radius would not be visible to the user.

And no separate widgets needed. The logic is in setting up the
animations, not in the animated widgets, and those animations are
shared.

justinmc committed:
k in constant names

justinmc committed:
Update references to zoom transition

justinmc committed:
Phase docs

justinmc committed:
Review and docs cleanup

justinmc committed:
Test fixed to use correct fallback transition.

Since I recently changed it from Zoom to FadeForwards.

justinmc commented:
@chunhtai for re-review. I've refactored a lot of the animation logic in _PredictiveBackSharedElementPageTransitionState using your ideas about ProxyAnimation. I have boiled it down to three Animations that drive all of the different changes:

 1. _animation, for things whose animations before and after the commit are unrelated. Goes 0-1 then 0-1 again.
 2. _bounceAnimation, for things that animate back to their original state after the commit. Goes 0-1 then 1-0.
 3. _commitAnimation, for things that only animate after the commit. Goes 0-0 then 0-1.

justinmc committed:
Fix fullscreen animation

justinmc committed:
Fix leak

chunhtai reviewed (commented):
mostly looks good, just some question around animation and observer

TheLastFlame commented on packages/flutter/lib/src/material/predictive_back_page_transitions_builder.dart:
```diff
@@ -159,128 +295,409 @@ class _PredictiveBackGestureDetectorState extends State<_PredictiveBackGestureDe
+                  _getYShiftPosition(MediaQuery.heightOf(context)),
+                ),
+            },
+            child: Opacity(
```
@chunhtai, this is necessary to animate the current route and the one below it at the same time. We need to get this knowledge both there and there

justinmc replied:
Here's a video of native. It's fast so it's hard to tell, but it does look like it's fading out to me.

https://github.com/user-attachments/assets/669825b0-3840-4eff-a449-5ee74884ffe6

chunhtai replied:
In that case should we lower the animation duration for commit animation? right now it was a bit slow, also I assume we will have delegate transition for the route beneath right?

TheLastFlame commented on packages/flutter/lib/src/widgets/binding.dart:
```diff
@@ -948,32 +947,33 @@ mixin WidgetsBinding
-  // The observer that is currently handling an active predictive back gesture.
-  WidgetsBindingObserver? _backGestureObserver;
+  // The observers that are currently handling an active predictive back gesture.
+  final List<WidgetsBindingObserver> _backGestureObservers = <WidgetsBindingObserver>[];
```
@chunhtai, this is necessary to animate the current route and the one below it at the same time. We need to get this properties both there and there

justinmc committed:
Formatted for analyzer

justinmc committed:
Comment explaining variable

justinmc committed:
Fix CurvedAnimations.

They weren't quite right after I fixed the leak in a previous commit. I
made some accidental logical change there.

chunhtai reviewed (approved):
LGTM

TheLastFlame commented:
Please note that the page should be scaled, not resized
The text should not rebuild
That's not the case now

https://github.com/user-attachments/assets/273513bc-3935-448d-922a-2a57ae79e8eb

justinmc commented:
> Please note that the page should be scaled, not resized The text should not rebuild That's not the case now

@TheLastFlame Thanks for reviewing, good catch. @maRci002 do you remember why you animated both the size of the page and the padding? I'm going to remove the padding animation and just go with scale only as long as it looks correct.

justinmc committed:
Remove padding animation

justinmc reviewed (approved):
LGTM :+1: 

I wrote most of this PR but the bot requires 2 approvals :smile: .

maRci002 commented:
Thank you so much, @justinmc, for your work the PR looks good to me.
I did notice a small bug: when you open the page and the animation is still ongoing, if you immediately perform the back gesture, it's not recognized. However, this can be fixed later.

https://github.com/user-attachments/assets/fc65f481-bad4-4a2e-8a55-413ce6c4b32b





> @maRci002 do you remember why you animated both the size of the page and the padding?

The [spec](https://developer.android.com/design/ui/mobile/guides/patterns/predictive-back#motion-specs) says:

> Leave an 8 dp gap from the screen edge
> Limit y-shift so the surface never passes the 8 dp screen margin

but I think the animation looks good now.

TheLastFlame commented:
> I did notice a small bug: when you open the page and the animation is still ongoing, if you immediately perform the back gesture, it's not recognized. However, this can be fixed later.

I also recently noticed this issue while working on https://github.com/TheLastFlame/universal-back-gesture.

It's not only relevant for this predicative builder, but also for cupertino and all custom builders that rely on the page closing gesture. 
The idea is that flutter blocks bottom screen taps as long as popUserGestureInProgress == true.

I think we need to find this and remove it
